### ISSUE TITLE
Readonly Transaction being committed

### ISSUE THREAD

**Issue by mburak** [grails/grails-data-mapping#580]
Even when a controller method has the Transactional(readOnly = true) annotation, the object is being updated.

This is how the method looks like:

```
@Transactional(readOnly = true)
def validate(Club club) {
    club.validate()

    respond club
}
```

and doing this call:

```
echo '{"id": -6934245603440287743, "name": "FC Bayern Münchenn"}' | http PUT http://localhost:8080/neo4j-app/clubs/-6934245603440287743/validate
```
it updates the club's name even when the method is readonly.

It looks that it's not considering the @Transactional(readOnly = true) on binding, just it is when entering the method.

**Comment by graemerocher** [comment]
So there is no `readOnly` state in Neo4j transactions, the only thing we can do is roll back the transaction if there is a `readOnly` transaction defined

**Comment by mburak** [comment]
Aren't you already doing that? Because if I do the binding manually inside the method it works fine and it doesn't update the object.
The issue only happens when I leave the binding to Grails.

**Comment by graemerocher** [comment]
This is an issue with the binding logic. In order to pass `club` to method Grails will call `Club.get(params.id)`. This happens prior to the read-only transaction starting, hence it becomes writable. The binding logic needs to be made `@Transactional` aware, but that is a problem in Grails core not in GORM itself. /cc @jeffbrown

**Comment by jeffbrown** [comment]
@mburak Can you provide a simple sample app which shows one action that uses the auto-binding and one that uses manual binding which demonstrated different behaviors?  I am having trouble recreating the behavior with Hibernate and the latest 2.5.x code.

**Comment by mburak** [comment]
Sure @jeffbrown , you can download it from http://github.com/mburak/neo4j-app. There's two actions in ClubController. One is validate that has auto-binding and the other is validateManualBinding that uses manual binding.  I am testing it using httpie and doing this kind of call:  
```echo '{"id": -6934245603440287743, "name": "FC Bayern Münchenn"}' | http PUT http://localhost:8080/neo4j-app/clubs/-6934245603440287743/validate```
You need to replace the id with the one it generates for you for any club. Then when you do that call it will update the club's name. If you doing with validateManualBinding it won't.

**Comment by jeffbrown** [comment]
I am seeing different behavior with hibernate.  The project at https://github.com/jeffbrown/txdemo contains the following:

```
package demo

import grails.transaction.Transactional

@Transactional
class DemoController {

    def create(Person p) {
        p.save()
        respond p
    }

    def show(Person p) {
        respond p
    }

    def autoBind(Person p) {
        respond p
    }

    def manualBind() {
        Person p = Person.get(params.id)
        p.properties = request
        respond p
    }

    @Transactional(readOnly=true)
    def autoBindReadOnly(Person p) {
        respond p
    }

    @Transactional(readOnly=true)
    def manualBindReadOnly() {
        Person p = Person.get(params.id)
        p.properties = request
        respond p
    }
}
```

The read-only nature of the transactions appears to be respected:

```

~ $ curl http://localhost:8080/txdemo/demo/show/1.json
{"class":"demo.Person","id":1,"name":"Jeff"}
~ $ 
~ $ curl -X PUT -H "Content-Type: application/json" -d '{"name":"First New Name"}' http://localhost:8080/txdemo/demo/autoBind/1.json
{"class":"demo.Person","id":1,"name":"First New Name"}
~ $ 
~ $ curl http://localhost:8080/txdemo/demo/show/1.json
{"class":"demo.Person","id":1,"name":"First New Name"}
~ $ 
~ $ curl -X PUT -H "Content-Type: application/json" -d '{"name":"Second New Name"}' http://localhost:8080/txdemo/demo/manualBind/1.json
{"class":"demo.Person","id":1,"name":"Second New Name"}
~ $ 
~ $ curl http://localhost:8080/txdemo/demo/show/1.json
{"class":"demo.Person","id":1,"name":"Second New Name"}
~ $ 
~ $ curl -X PUT -H "Content-Type: application/json" -d '{"name":"Third New Name"}' http://localhost:8080/txdemo/demo/autoBindReadOnly/1.json
{"class":"demo.Person","id":1,"name":"Third New Name"}
~ $ 
~ $ curl http://localhost:8080/txdemo/demo/show/1.json
{"class":"demo.Person","id":1,"name":"Second New Name"}
~ $ 
~ $ curl -X PUT -H "Content-Type: application/json" -d '{"name":"Fourth New Name"}' http://localhost:8080/txdemo/demo/manualBindReadOnly/1.json
{"class":"demo.Person","id":1,"name":"Fourth New Name"}
~ $ 
~ $ curl http://localhost:8080/txdemo/demo/show/1.json
{"class":"demo.Person","id":1,"name":"Second New Name"}
~ $ 
```

The project contains the following test which passes:

```
package demo

import grails.test.mixin.TestFor
import grails.test.mixin.TestMixin
import grails.test.mixin.hibernate.HibernateTestMixin
import spock.lang.Specification
import spock.lang.Stepwise

/**
 * See the API for {@link grails.test.mixin.web.ControllerUnitTestMixin} for usage instructions
 */
@TestFor(DemoController)
@TestMixin(HibernateTestMixin)
@Stepwise
class DemoControllerSpec extends Specification {

    void setupSpec() {
        hibernateDomain([Person])
    }

    void "test creating person"() {
        when:
        request.addHeader 'Accept', 'application/json'
        request.method = 'POST'
        request.json = '{"name":"original name"}'
        controller.create()

        then:
        response.status == 200
        response.text.contains '"original name"'
    }

    void "test retrieving person"() {
        when:
        request.addHeader 'Accept', 'application/json'
        params.id = 1
        controller.show()

        then:
        response.status == 200
        response.text.contains '"original name"'
    }

    void "test auto bind"() {
        when:
        request.addHeader 'Accept', 'application/json'
        request.method = 'PUT'
        request.json = '{"name":"first new name"}'
        params.id = 1
        controller.autoBind()

        then:
        response.status == 200
        response.text.contains '"first new name"'
    }

    void "test retrieving person after auto bind"() {
        when:
        request.addHeader 'Accept', 'application/json'
        params.id = 1
        controller.show()

        then:
        response.status == 200
        response.text.contains '"first new name"'
    }

    void "test manual bind"() {
        when:
        request.addHeader 'Accept', 'application/json'
        request.method = 'PUT'
        request.json = '{"name":"second new name"}'
        params.id = 1
        controller.manualBind()

        then:
        response.status == 200
        response.text.contains '"second new name"'
    }

    void "test retrieving person after manual bind"() {
        when:
        request.addHeader 'Accept', 'application/json'
        params.id = 1
        controller.show()

        then:
        response.status == 200
        response.text.contains '"second new name"'
    }

    void "test manual bind read only"() {
        when:
        request.addHeader 'Accept', 'application/json'
        request.method = 'PUT'
        request.json = '{"name":"third new name"}'
        params.id = 1
        controller.manualBindReadOnly()

        then:
        response.status == 200
        response.text.contains '"third new name"'
    }

    void "test retrieving person after manual bind read only"() {
        when:
        request.addHeader 'Accept', 'application/json'
        params.id = 1
        controller.show()

        then:
        response.status == 200
        response.text.contains '"second new name"'
    }

    void "test auto bind read only"() {
        when:
        request.addHeader 'Accept', 'application/json'
        request.method = 'PUT'
        request.json = '{"name":"fourth new name"}'
        params.id = 1
        controller.autoBindReadOnly()

        then:
        response.status == 200
        response.text.contains '"fourth new name"'
    }

    void "test retrieving person after auto bind read only"() {
        when:
        request.addHeader 'Accept', 'application/json'
        params.id = 1
        controller.show()

        then:
        response.status == 200
        response.text.contains '"second new name"'
    }
}
```

We will figure out why this behaves differently for neo4j.

**Comment by jeffbrown** [comment]
I thought the issue here might be the fact that the query was happening in the no-arg method that we generate at compile time and then passing that object into the method that is defined in the source code.  I have tested with a version of our code that moves the query into the original method, and I am seeing the same behavior:

```
neo4j-app master $ curl http://localhost:8080/neo4j-app/clubs/-6663955238840057856
{"class":"app.Club","id":-6663955238840057856,"country":null,"dateCreated":"2016-02-24T19:10:37Z","lastUpdated":"2016-02-24T19:10:37Z","league":{"class":"app.League","id":-6663955238840066048},"name":"Bayer 04 Leverkusen","players":[{"class":"app.Player","id":-6663955238840057855},{"class":"app.Player","id":-6663955238840057854},{"class":"app.Player","id":-6663955238840057853}]}neo4j-app master $ 
neo4j-app master $ 
neo4j-app master $ 
neo4j-app master $ curl -X PUT -H "Content-Type: application/json" -d '{"name":"First Update"}' http://localhost:8080/neo4j-app/clubs/-6663955238840057856/validateManualBinding
{"class":"app.Club","id":-6663955238840057856,"country":null,"dateCreated":"2016-02-24T19:10:37Z","lastUpdated":"2016-02-24T19:10:37Z","league":{"class":"app.League","id":-6663955238840066048},"name":"First Update","players":[{"class":"app.Player","id":-6663955238840057855},{"class":"app.Player","id":-6663955238840057854},{"class":"app.Player","id":-6663955238840057853}]}neo4j-app master $ 
neo4j-app master $ 
neo4j-app master $ curl http://localhost:8080/neo4j-app/clubs/-6663955238840057856
{"class":"app.Club","id":-6663955238840057856,"country":null,"dateCreated":"2016-02-24T19:10:37Z","lastUpdated":"2016-02-24T19:10:37Z","league":{"class":"app.League","id":-6663955238840066048},"name":"Bayer 04 Leverkusen","players":[{"class":"app.Player","id":-6663955238840057855},{"class":"app.Player","id":-6663955238840057854},{"class":"app.Player","id":-6663955238840057853}]}neo4j-app master $ 
neo4j-app master $ 
neo4j-app master $ curl -X PUT -H "Content-Type: application/json" -d '{"name":"Second Update"}' http://localhost:8080/neo4j-app/clubs/-6663955238840057856/validate
{"class":"app.Club","id":-6663955238840057856,"country":null,"dateCreated":"2016-02-24T19:10:37Z","lastUpdated":"2016-02-24T19:10:37Z","league":{"class":"app.League","id":-6663955238840066048},"name":"Second Update","players":[{"class":"app.Player","id":-6663955238840057855},{"class":"app.Player","id":-6663955238840057853},{"class":"app.Player","id":-6663955238840057854}]}neo4j-app master $ 
neo4j-app master $ 
neo4j-app master $ curl http://localhost:8080/neo4j-app/clubs/-6663955238840057856
{"class":"app.Club","id":-6663955238840057856,"country":null,"dateCreated":"2016-02-24T19:10:37Z","lastUpdated":"2016-02-24T19:10:37Z","league":{"class":"app.League","id":-6663955238840066048},"name":"Second Update","players":[{"class":"app.Player","id":-6663955238840057855},{"class":"app.Player","id":-6663955238840057853},{"class":"app.Player","id":-6663955238840057854}]}
```

Still investigating.

**Comment by jeffbrown** [comment]
I think I see the issue now.  Will work toward a solution.

**Comment by jeffbrown** [comment]
@mburak This may be resolved in the latest 2.5.4.BUILD-SNAPSHOT.  Are you setup to be able to test that and provide feedback?

**Comment by mburak** [comment]
Sure, I'll try.

**Comment by jeffbrown** [comment]
@mburak Great.  Please let me know how that goes.  I tested with your sample app and the behavior looks right.  If you can confirm that you are seeing same, we can call it good and ship it.

Thanks for the feedback.

**Comment by mburak** [comment]
Hi @jeffbrown, I was about to test this but using sdk I can't find the snapshot. I tried this command:
```sdk install grails 2.5.4.BUILD-SNAPSHOT``` 
and also tried this one: 
```sdk install grails 2.5.4-BUILD-SNAPSHOT```
The response is always the same:
```Stop! 2.5.4-BUILD-SNAPSHOT is not a valid grails version.```

How should I install it?
Thanks.

**Comment by jeffbrown** [comment]
You can do something like this:

```
cd /somedirectory
git clone git@github.com:grails/grails-core.git
cd grails-core
./gradlew install
sdk install grails dev /somedirectory/grails-core
sdk use grails dev
```
Note that `dev` is just a label and could be just about anything.

**Comment by jeffbrown** [comment]
You can speed the build up by using `./gradlew install -x javadoc` instead of `./gradlew install`.

**Comment by jeffbrown** [comment]
There are some relevant notes at https://github.com/grails/grails-core/wiki/Building-A-Local-Grails-Installation.

**Comment by mburak** [comment]
Thank you Jeff. It looks like working. 
The only problem I see is that I had to put the Transactional annotation at method level to make it work.
If I have the annotation at class level, it doesn’t consider it. Is this something you can fix?

Thanks.

**Comment by jeffbrown** [comment]
> The only problem I see is that I had to put the Transactional annotation at method level to make it work.

Yep.  You are right.  This is a mistake.  Let me teach the transformation to handle that better.  Good catch!

**Comment by jeffbrown** [comment]
@mburak I think I may have resolved the issue now but I have not yet been able to verify that against your sample project.  I can do that later today but if you get a chance to test that yourself, please provide feedback here.

I think you probably figured it out but I failed to mention above that in order to test your changes locally, you will want to be working with the `2.5.x` branch.

**Comment by mburak** [comment]
Yes Jeff, it’s working fine.
And yeah, I’m working with the 2.5.x branch.

When do you think we can have a release?

Thanks.

**Comment by jeffbrown** [comment]
>Yes Jeff, it’s working fine.

Great!

> When do you think we can have a release?

Probably tomorrow.  I will keep you posted.

Thanks for the feedback.

**Comment by AnjanaBangalore** [comment]
We are using grail 2.4.4 and we are facing same issue where in readonly transaction is auto committed. We are not ready yet to upgrade grails due to time factor. Is there a way to get patch for this fix in 2.4.4 this will help us resolve the issue in production.

**Comment by graemerocher** [comment]
Nope, you are welcome to go through the support channels however https://grails.org/support.html
